### PR TITLE
Particle injection

### PR DESCRIPTION
**Pull request by egpuckett**
Added the ability to inject particles at a time step N specified in the parameters.prm file and continue to inject particles every N time steps, or just the one time step N depending on a Boolean flag. The user can also choose to generate particles at time T = 0 or not. The positions of the particles are determined by the routines uniform_box.cc, uniform_radial.cc, ascii_file.cc, but not random_uniform.cc and probability_density_function.cc.

### CODE REVIEW

**Comment by gassmoeller** [comment]
Hi @hlokavarapu, Hi Gerry. Thanks for this PR. I like the functionality, but of course I have some functional and stylistic comments :smile:. First some general things:

- As a rule of thumb we use CamelCase only for namespaces and class names, but not for variables and function names. For those we prefer lower case divided by underscored (is_time_to_generate_particles() instead of isTimeToGenerateParticles()). Please take a look through your code and change where appropriate.

- You have made Generator::Interface responsible for figuring out at which time particle generation is needed and ask it every timestep in World::advance_timestep() if it is time to do so. On the other hand in the already implemented structure Simulator is responsible for calling the particle generation at the correct time (at model start), and the generator is just accepting this judgement, because it only knows *how* to generate particles, not *when*. I am aware that Simulator is not the perfect place to do it either, but I would strongly prefer to not split this responsibility into several places. Probably World would be the place to put it in. I could imagine a new function World::update() that is called from Simulator at every timestep like for the other plugins (see e.g. core.cc::786). Then this function could figure out if it is time to generate new particles and call the generator if appropriate. This way we would also not pollute advance_timestep() with other functionalities besides the main purpose of this function. The necessary input parameters could also move to World.

I will go through the code and mark up smaller things, but we should discuss my second point above, because it will be a decision that might have serious consequences for future changes.

File: cookbooks/particle_injection.prm [hunk 1]
```diff
@@ -0,0 +1,247 @@
+# At the top, we define the number of space dimensions we would like to
```
Comment (gassmoeller) [inline cookbooks/particle_injection.prm]: Please go through the file and remove unnecessary comment lines. Also add a small sentence at the beginning what this test is about. Not super important, but nicer if somebody needs to look into this test in the future.

File: cookbooks/particle_injection.prm [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+# and we choose to compute velocity and temperature statistics,
+# statistics about the heat flux through the boundaries of the
+# domain, and to generate graphical output files for later
+# visualization. 
+
+subsection Postprocess
+   set List of postprocessors = basic statistics, visualization, tracers, heat flux statistics, velocity statistics
+#  set List of postprocessors = basic statistics, visualization, heat flux statistics, velocity statistics
+#  set List of postprocessors = basic statistics, heat flux statistics, velocity statistics
+#  set List of postprocessors = basic statistics, heat flux statistics, velocity statistics, tracers
```
Comment (gassmoeller) [inline cookbooks/particle_injection.prm]: Some unnecessary lines here. Also we usually deactivate visualization output whenever possible for tests. Not important if you do not check in visualization test results, but if others copy this test they will also end up with visualization output enabled.

File: cookbooks/particle_injection.prm [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    set Time between data output = 0
+    set Data output format = vtu
+
+    set Particle generator name = uniform box
+    set Integration scheme = rk2
+
+    set List of tracer properties = initial position, position, velocity
+
+    subsection Generator
+      # Prevent particles from being placed on the boundary layer within a
```
Comment (gassmoeller) [inline cookbooks/particle_injection.prm]: Which generator is actually used for this test? Supposedly the default one? Could you still add that line here, in case the default gets changed at some point?

File: include/aspect/particle/generator/interface.h [hunk 4]
```diff
@@ -126,6 +126,21 @@ namespace aspect
           void
           parse_parameters (ParameterHandler &prm);
 
+          void
+          parse_global_parameters (ParameterHandler &prm);
```
Comment (gassmoeller) [inline include/aspect/particle/generator/interface.h]: I think there is no need for this additional function. If there are parameters that should be read by the interface you can put them into the default implementation of `parse_parameters` and call that function from derived plugins.

File: include/aspect/particle/generator/interface.h [hunk 5]
```diff
@@ -126,6 +126,21 @@ namespace aspect
...
+
+          bool
+          isTimeToGenerateParticles(unsigned int timeStepNumber);
```
Comment (gassmoeller) [inline include/aspect/particle/generator/interface.h]: See my general comment about names. Also this could move to World if you agree with my second point (I am open for discussion about that though).

File: include/aspect/particle/generator/interface.h [hunk 6]
```diff
@@ -126,6 +126,21 @@ namespace aspect
...
+
+          bool
+          getGenerateParticlesInitially();
+
+          unsigned int
+          getParticleIdx();
```
Comment (gassmoeller) [inline include/aspect/particle/generator/interface.h]: This is in principle a good property to know about. However I have the feeling that it somehow duplicates information that is already stored in World::next_free_particle_index. Could you take a look and think about if it is possible to remove this duplication? Maybe there is not, but I am not sure, and storing duplicated information always leads to trouble.

File: source/particle/generator/interface.cc [hunk 7]
```diff
@@ -136,14 +137,101 @@ namespace aspect
...
+            }
+            prm.leave_subsection();
+          }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+
+        if (repeatedlyGenerateParticles)
+          {
+            AssertThrow((name != "probability density function") && (name != "random uniform"),
```
Comment (gassmoeller) [inline source/particle/generator/interface.cc]: You should not rely on which plugins are implemented right now. If someone adds a plugin, but forgets to add it to this list your check will not work anymore. Better check if one of the supported plugins is selected.

File: source/particle/world.cc [hunk 8]
```diff
@@ -1127,6 +1131,13 @@ namespace aspect
     void
     World<dim>::advance_timestep()
     {
+      // Generate more particles
+      if (generator->isTimeToGenerateParticles(this->get_timestep_number()))
+        {
+          generate_particles();
+          initialize_particles();
+        }
```
Comment (gassmoeller) [inline source/particle/world.cc]: I would like to find a different place for this piece of code (e.g. and update() function). One reason is that I am considering calling this function multiple time per timesteps to iterate over particle advection and Stokes solution, and in this case I do not want to generate particles in every iteration.

File: cookbooks/particle_injection.prm (hunk 2 above)
Comment (bangerth) [inline cookbooks/particle_injection.prm]: This isn't actually a test, but in the `cookbooks/` directory.

File: source/particle/world.cc [hunk 9]
```diff
@@ -1215,6 +1226,7 @@ namespace aspect
       }
       prm.leave_subsection ();
 
+      Generator::Interface<dim>::declare_parameters(prm);
```
Comment (gassmoeller) [inline source/particle/world.cc]: This call should probably come from the child classes of the Interface that need these parameters. Here it looks a bit out of place.

File: source/particle/world.cc [hunk 10]
```diff
@@ -1276,6 +1288,7 @@ namespace aspect
       generator.reset(Generator::create_particle_generator<dim> (prm));
       if (SimulatorAccess<dim> *sim = dynamic_cast<SimulatorAccess<dim>*>(generator.get()))
         sim->initialize_simulator (this->get_simulator());
+      generator->parse_global_parameters(prm);
```
Comment (gassmoeller) [inline source/particle/world.cc]: Same here, the call to `Interface::parse_parameters` should likely come from the derived classes via the call to `generator->parse_parameters(prm);` below

File: cookbooks/particle_injection.prm (hunk 2 above)
Comment (gassmoeller) [inline cookbooks/particle_injection.prm]: Oh, missed that completely. @hlokavarapu could you also add a test, and if you want to write a cookbook maybe put it in here as well, or make a separate PR for that?

Comment (egpuckett) [inline cookbooks/particle_injection.prm]: @gassmoeller @bangerth Would you please give us some guidance on what constitutes a good test.  For example we might want to check if the particles are injected at the correct place after say five time steps.  But the particle positions would be in an ascii file or a *.vtu file.  is it acceptable to do a diff on an ascii file? What other things might we test for, and would we typically design the text for a specific number of time steps?

Comment (bangerth) [inline cookbooks/particle_injection.prm]: A test really only needs to run a particular piece of code in a way that you can check once, when you write the test, that it does the correct thing and so that if it is broken, it produces different output. The testcase doesn't have to realistic at all; it just needs to produce different output if someone changes the code in incorrect ways.

Many of the tests we have work on 2x2 grids or something similarly pointless. But that's good enough for the purpose of a test :-)

**Comment by bangerth** [comment]
This patch is by now 8 years old, and essentially none of it will still apply given that the particle infrastructure has completely changed. Let's close this. If someone wants to re-consider this kind of functionality, the implementation will have to start from scratch anyway.
